### PR TITLE
no paginating to find groups

### PR DESCRIPTION
**Pull request by hopkiw**
* replace FindGroup paginating function with GetGroupBy{Name,GID}
* split getgrby function to match

### CODE REVIEW

File: src/oslogin_utils.cc [hunk 1]
```diff
@@ -813,80 +821,83 @@ bool FindGroup(struct group* result, BufferManager* buf, int* errnop) {
...
   long http_code;
   string pageToken = "";
 
-  do {
-    url.str("");
-    url << kMetadataServerUrl << "groups?email=" << email;
-    if (pageToken != "")
-      url << "&pagetoken=" << pageToken;
+  url.str("");
+  char gidstring[65];
```
Comment (illfelder) [inline src/oslogin_utils.cc:875]: Why is 65 hardcoded?

File: src/oslogin_utils.cc [hunk 2]
```diff
@@ -813,80 +821,83 @@ bool FindGroup(struct group* result, BufferManager* buf, int* errnop) {
...
+  if (sprintf(gidstring,"%d",gid) < 1) {
```
Comment (illfelder) [inline src/oslogin_utils.cc:876]: I think there should be spaces after the commas.

File: src/oslogin_utils.cc [hunk 3]
```diff
@@ -781,25 +783,31 @@ bool ParseJsonToChallenges(const string& json, std::vector<Challenge>* challenge
...
 
-// TODO: this function reads all groups comparing names or gids; it should be
-// replaced by groups?groupname= lookup when this is available.
-bool FindGroup(struct group* result, BufferManager* buf, int* errnop) {
-  if (result->gr_name == NULL && result->gr_gid == 0) {
-    // Nobody told me what to find.
+bool GetGroupsForUser(string username, std::vector<Group>* groups, int* errnop) {
+  string response;
+  if (!(GetUser(username, &response))) {
+    DEBUG("GetGroupsForUser: !GetUser\n");
```
Comment (illfelder) [inline src/oslogin_utils.cc:789]: This intentional?

File: src/oslogin_utils.cc (hunk 1 above)
Comment (hopkiw) [reply src/oslogin_utils.cc:875]: it is the maximum string length of a 64-bit integer.

File: src/oslogin_utils.cc (hunk 3 above)
Comment (hopkiw) [reply src/oslogin_utils.cc:789]: is what intentional?

**Comment by winterfreshx** [comment]
Have we added a test case to simulate the large group and user problem that we found?

**Comment by hopkiw** [comment]
> Have we added a test case to simulate the large group and user problem that we found?

will add regression tests in our e2e test suite to cover these. the functions that changed make HTTP requests so tough to unit test them without implementing an abstraction layer just for that purpose.

**Pull request merged**